File: src/19-challenge/solution.js
/**
 * @param {string[][]} routes - Array of [origin, destination] pairs
 * @returns {string[]} The reconstructed route
 */
export default function fn(routes) {
	const reconstructedRoute = [...routes[0]]

	let destination = reconstructedRoute.at(-1)

	while (destination) {
		const tripSegment = routes.find(([origin]) => origin === destination)
		if (!tripSegment) break

		destination = tripSegment[1]
		reconstructedRoute.push(destination)
	}

	return reconstructedRoute
}

File: src/19-challenge/solution.py
def fn(routes: list[list[str]]) -> list[str]:
    reconstructed_route = [*routes[0]]

    destination = reconstructed_route[-1]

    while destination:
        trip_segment: list[str] = []

        for segment in routes:
            if segment[0] == destination:
                trip_segment = segment
                break

        if not trip_segment:
            break

        destination = trip_segment[1]
        reconstructed_route.append(destination)

    return reconstructed_route
